src/erlcloud_ddb_streams.erl

```diff
@@ -0,0 +1,907 @@
+%%
+%% @end
+%%------------------------------------------------------------------------------
+-spec list_streams(list_streams_opts(), aws_config()) -> list_streams_return().
+list_streams(Opts, Config) ->
+    {AwsOpts, DdbStreamsOpts} = opts(list_streams_opts(), Opts),
+    Return = request(
+               Config,
+               "DynamoDBStreams_20120810.ListStreams",
+               AwsOpts),
+    out(Return,
+        fun(Json, UOpts) -> undynamize_record(list_streams_record(), Json, UOpts) end,
+        DdbStreamsOpts, #ddb_streams_list_streams.streams).
+
+%%%------------------------------------------------------------------------------
+%%% Request
+%%%------------------------------------------------------------------------------
+
+-type operation() :: string().
+-type json_return() :: {ok, json_term()} | {error, term()}.
+
+-spec request(aws_config(), operation(), json_term()) -> json_return().
+request(Config, Operation, Json) ->
+    case erlcloud_aws:update_config(Config) of
+        {ok, Config2} ->
+            request2(Config2, Operation, Json);
+        {error, _} = Error ->
+            Error
+    end.
+
+-spec request2(aws_config(), operation(), json_term()) -> json_return().
+request2(Config, Operation, Json) ->
+    Body = case Json of
+               [] -> <<"{}">>;
+               _ -> jsx:encode(Json)
+           end,
+    Headers = headers(Config, Operation, Body),
+    Request = #aws_request{service = ddb_streams,
+                           uri = uri(Config),
+                           method = post,
+                           request_headers = Headers,
+                           request_body = Body},
+    request_to_return(erlcloud_retry:request(Config, Request, fun result_fun/1)).
+
+-spec request_to_return(#aws_request{}) -> json_return().
+request_to_return(#aws_request{response_type = ok,
+                               response_body = Body}) ->
+    %% TODO check crc
+    {ok, jsx:decode(Body)};
+request_to_return(#aws_request{response_type = error,
+                               error_type = aws,
+                               httpc_error_reason = Reason}) ->
+    {error, Reason};
+request_to_return(#aws_request{response_type = error,
+                               error_type = httpc,
+                               httpc_error_reason = Reason}) ->
+    {error, Reason}.
+
+-spec result_fun(#aws_request{}) -> #aws_request{}.
+result_fun(#aws_request{response_type = ok} = Request) ->
+    Request;
+result_fun(#aws_request{response_type = error,
+                        error_type = aws,
+                        response_status = Status,
+                        response_status_line = StatusLine,
+                        response_body = Body} = Request) ->
+    %% == IMPLEMENTATION NOTES ==
+    %% We store the error reason in `httpc_error_reason` for now,
+    %% this may be changed at any time later
+    Request2 = Request#aws_request{
+        httpc_error_reason = {http_error, Status, StatusLine, Body}},
```

Comment:
Seems it's a better way, I'll have a detailed think and modify it if necessary a little later.